danceradvent/public/articles/2015/3-medium-scale-part-1.pod

```diff
@@ -0,0 +1,100 @@
+"uninteresting" here because we recommend that you leave these files
+as-is. None of your app-specific Perl code should go in these files.
+
+We're also ignoring C<Makefile.PL> since that isn't part of the app
+proper.
+
+That only leaves C<lib/App.pm>. For small-scale apps, simply piling all
+of your Dancer-related app code into this file is just fine. Up to about
+one or two thousand lines of code, you can just use it as-is, for the
+same reason that a thousand-line monolithic Perl script is fine in other
+contexts, too.
+
+=head2 Extending the Stock Scheme
+
+One of the nicest things about Dancer is that the generated startup code
+adds the app's C<lib> directory to the Perl module path. That leads to
+an obvious solution to the problem of too much code in a single file:
+break it out into other C<*.pm> files in the C<lib> directory. But how,
+exactly?
+
+I recommend following the example of CPAN, with a structure like the
+following:
+
+ lib/App.pm                 # your app's main module
+ lib/App/MajorFeature.pm    # implementation of a major app feature
+ lib/App/OtherFeature.pm    # as many feature impl files as you need
+ lib/App/API.pm             # if your app has a REST API, put it here
+ lib/App/Const.pm           # maybe you have some app-wide constants
+ lib/App/Utility.pm         # every big app has a "junk drawer" module
+
+Then in C<lib/App.pm>, you C<use> all of these modules:
+
+ use App::MajorFeature;
+ use App::OtherFeature;
+ use App::API;
+ use App::Const;
+ use App::Utility;
+
+How you design these modules is up to you. The
+L<standard rules|http://stackoverflow.com/questions/410359/how-can-i-learn-to-write-well-structured-programs-in-perl>
+for well-structured Perl application design apply.
+
+There is one Dancer2-specific thing to beware of here, though: you must
+L<set the C<appname> property on the sub-modules|http://advent.perldancer.org/2014/10>. If you
+don't, each sub-module will be considered a separate Dancer2
+application, which puts unwanted barriers between the modules.
+
+(Dancer1 didn't have this requirement because it didn't have any
+concept of multiple apps within a single Dancer instance.)
+
+When it comes time to break C<lib/App.pm> up into multiple modules, I
+recommend that you move almost all Perl code into one of these modules,
+leaving only application initialization code in the top-level module.
+
+(This, by the way, is why we don't need to modify C<bin/app.psgi>: since
+it pretty much just loads Dancer and then calls C<lib/App.pm>, there is
+no point in splitting app initialization code across the two files. Let
+C<bin/app.psgi> remain a tiny little glue file that exists merely to
+make L<PSGI|http://plackperl.org/> happy.)
+
+Your initial refactoring of C<lib/App.pm> will likely leave all the
+route definitions behind, which is fine because refactoring the routes
+is the subject of the
+L<next article in this series|http://advent.perldancer.org/2015/4>.
```

Comment:
We should add this link once the articles are published, so it works. At first it won't work, because the next article isn't published yet.